### PR TITLE
feat(upgrade): implement dependency tree

### PR DESCRIPTION
**Pull request by Elsie19**
## Purpose

Packages like [rhino-core](https://github.com/pacstall/pacstall-programs/blob/master/packages/rhino-core/rhino-core.pacscript) and [linux-kernel](https://github.com/pacstall/pacstall-programs/blob/master/packages/linux-kernel/linux-kernel.pacscript) are metapackages that hold calls to install other pacscripts; and due to how the pacdeps function works, a specific order to install those pacscripts. If the dependencies of these metapackages are installed before they are called for the proper order, installs may fail. The dependency tree aims to smooth this out.

## Approach

Associative array magic 🪄. Basically we yeet out pacdeps from the update equation, because `-I` checks for updates of pacdeps, then we just order everything.

## Other

> **Note**
> Order printing is not correct, but it updates in the correct order. May or may not work on that.

## Checklist

- [x] I confirm that I have read the [contributing guidelines](https://github.com/pacstall/pacstall/blob/develop/CONTRIBUTING.md), and this pull request is abiding by all the clauses stated in the guideline.

### CODE REVIEW

**Comment by Henryws** [comment]
Currently an issue reported by Oren that if a pacdep is updated, but not the parent, the pacdep won't update.

**Comment by oklopfer** [comment]
> Currently an issue reported by Oren that if a pacdep is updated, but not the parent, the pacdep won't update because it's stripped from the upgrade process.

Clarification: as an example, rhino-pkg-git is a pacdep of rhino-core. With the dep-tree in its current state, `pacstall -Up` will only find if rhino-core updates; if rhino-pkg-git updates separately from the core, it is not found by the `Up` process.

**Comment by Henryws** [comment]
If the pacdep's dependee (parent package) is not in the upgrade, we should upgrade the pacdep, but if the parent package is in the upgrade, disregard it because `-I` takes care of pacdep upgrades.

File: misc/scripts/upgrade.sh [hunk 1]
```diff
@@ -22,23 +22,39 @@
...
 # Get the list of the installed packages
-mapfile -t list < <(pacstall -L)
+mapfile -t list < <(pacstall -L | awk NF)
+if ((${#list[@]} == 0)); then
+    fancy_message info "Nothing to upgrade"
+    return 0
+fi
+fancy_message info "Building dependency tree"
+dep_tree.loop_traits update_order "${list[@]}"
+dep_tree.trim_pacdeps update_order
```
Comment (oklopfer) [inline misc/scripts/upgrade.sh:51]: get rid of him

**Review by oklopfer — COMMENTED** [review]
good work, soldier. LGTM 🛸

**Review by oklopfer — APPROVED** [review]

**Pull request merged**